### PR TITLE
[extension/opampextension] reports effective config

### LINKED ISSUE #27293: [extension/opampextension] Subscribe to be notified of Collector's effective config
Implement `extension.NotifyConfig` to be notified of the Collector's effective config and report it to the OpAMP server.

### PR DESCRIPTION
**Pull request by haoqixu**
**Link to tracking Issue:** resolve #27293

**Testing:**
unittest
manual test with opamp-go example:
* collector version: modified otelcontribcol including opampextension
* collector config: 
```yaml
receivers:
  # otlp:
  #   protocols:
  #     grpc:
  filelog:
    include:
      - /tmp/xxxxx.log

exporters:
  debug:
    #verbosity: debug

extensions:
  opamp:
    server:
      ws:
        endpoint: ws://127.0.0.1:4320/v1/opamp
        tls:
          insecure_skip_verify: true

service:
  pipelines:
    logs:
      receivers: [filelog]
      exporters: [debug]
  telemetry:
    logs:
      level: info
  extensions: [opamp]
```
* server ui: Screenshot 2023-11-28 at 22-14-59 OpAMP Server

**Documentation:** README.md

### CODE REVIEW

**Review by JaredTan95 — APPROVED** [review]

File: .chloggen/feat-27293-reports-effective-config.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: opampextension
```
Comment (JaredTan95) [inline .chloggen/feat-27293-reports-effective-config.yaml:7]:
```suggestion
component: extension/opampextension
```

**Comment by JaredTan95** [comment]
Sorry, I planned to modify the review comment again, but I chose the wrong `hide comments` option(This comment was marked as abuse.)~

**Review by tigrannajaryan — DISMISSED** [review]

**Comment by tigrannajaryan** [comment]
> manual test with opamp-go example

Can you please attach a screenshot showing what it looks like in the server UI?

File: extension/opampextension/README.md [hunk 2]
```diff
@@ -27,6 +27,7 @@ The following settings are optional:
 - `instance_uid`: A ULID formatted as a 26 character string in canonical
   representation. Auto-generated on start if missing. Setting this ensures the
   instance UID remains constant across process restarts.
+- `reports_effective_config`: Whether to enable the OpAMP ReportsEffectiveConfig capability. Default is `true`.
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply extension/opampextension/README.md:30]: +1. That's also what the [design doc](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/cmd/opampsupervisor/specification/README.md#supervisor-configuration) says.

File: extension/opampextension/opamp_agent.go [hunk 3]
```diff
@@ -121,6 +110,22 @@ func (o *opampAgent) Shutdown(ctx context.Context) error {
...
+		return o.opampClient.UpdateEffectiveConfig(ctx)
+	}
+	return nil
+}
+
+func (o *opampAgent) updateEffectiveConfig(conf *confmap.Conf) {
+	o.eclk.Lock()
+	defer o.eclk.Unlock()
+
+	o.effectiveConfig = conf
```
[the remark this answers is not in the document]
Comment (haoqixu) [reply extension/opampextension/opamp_agent.go:125]: resolved in open-telemetry/opentelemetry-collector#9005

**Review by tigrannajaryan — APPROVED** [review]

**Review by evan-bradley — APPROVED** [review]

**Review by portertech — APPROVED** [review]

**Pull request merged**